Optimize file management for certificate previews

BarryPaneer opened the pull request:
# Task
 - https://foundever.atlassian.net/browse/TRIB-1486

# What the PR does
 - Support to compress and upload zip file including pdf files loaded from AWS/S3(Not Local disk) when exporting certificate of leaners.
 - COPY_TO_WEB_ROOT set to false by remove loading from JSON configuration file.
 - Generate a PNG file only for example certificates, not for real learner certificates. We use FLAG [example_certificate](https://github.com/Learningtribes/platform/blob/d8142a402b0599dcffadd1c04e328c981a06e234/lms/djangoapps/certificates/queue.py#L642)
 - Delete all local files in `downloads + certs` after **a Date** && Delete .PNG files associated with learner certificates on S3 by command 🏃  `/edx/app/certs/venvs/certs/bin/python /edx/app/certs/certificates/cleaner.py --target_type=[s3|dfs] --dryrun=[true|false]`
   - The date: We remove DFS files after the date 📅  `2020-2-12` when we started to upload files to S3 👇 . We can use `2020-2-15` as a conservative way : 
```
mysql> select created_date, modified_date, download_uuid, download_url from certificates_generatedcertificate where download_url is not null && download_url <> '' order by id asc limit 30;
+----------------------------+----------------------------+----------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| created_date               | modified_date              | download_uuid                    | download_url                                                                                                                                                                      |
+----------------------------+----------------------------+----------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 2018-03-22 17:14:43.088133 | 2019-12-09 11:35:39.738654 | fc242fb7179548e7848e2666d3e33768 | /downloads/fc242fb7179548e7848e2666d3e33768/nanglo924554_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                                |
| 2018-04-03 20:56:36.615797 | 2019-12-09 11:35:39.797760 | 3e9612b1f8c14491a8c945c2a30f124e | /downloads/3e9612b1f8c14491a8c945c2a30f124e/cmartin626800_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                               |
| 2018-04-17 18:10:14.838810 | 2019-12-09 11:35:39.824416 | 08d0a281ad5b428183694800538004c6 | /downloads/08d0a281ad5b428183694800538004c6/dsantiago869176_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                             |
| 2018-04-23 11:29:33.968943 | 2019-12-09 11:35:39.851691 | 0a35939bedf848ddafd5678f7d0cfdd7 | /downloads/0a35939bedf848ddafd5678f7d0cfdd7/sbosaing618610_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                              |
| 2018-05-24 15:49:48.284675 | 2019-12-09 11:35:39.909666 | 3f1e88a82fa847dfb4d72e0b8881f9ed | /downloads/3f1e88a82fa847dfb4d72e0b8881f9ed/jsingson1028990_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                             |
| 2018-06-27 20:58:47.737578 | 2019-12-09 11:35:40.069095 | b2e4915e90b74115802011e8cfd3bd05 | /downloads/b2e4915e90b74115802011e8cfd3bd05/mcastillo878685_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                             |
| 2018-06-28 13:24:20.049643 | 2019-12-09 11:35:40.097304 | d3ef4cbd4a7f4083a583fa220cb0a4b8 | /downloads/d3ef4cbd4a7f4083a583fa220cb0a4b8/esoriano618828_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                              |
| 2018-07-31 14:13:17.222006 | 2019-12-09 11:35:40.123063 | 84fbc9bbcd944661b3aee4eabc1bbf27 | /downloads/84fbc9bbcd944661b3aee4eabc1bbf27/iradcliffe746852_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                            |
| 2018-08-01 17:01:22.287025 | 2019-12-09 11:35:40.153232 | 10645a5f7a2d41288819673167c6190b | /downloads/10645a5f7a2d41288819673167c6190b/ldelcastillo588354_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                          |
| 2018-08-06 16:46:02.266141 | 2019-12-09 11:35:40.169276 | 841e41174d244051b0ef6cdbf3e74dc1 | /downloads/841e41174d244051b0ef6cdbf3e74dc1/jdiones567336_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                               |
| 2018-09-05 00:42:18.786082 | 2019-12-09 11:35:40.255069 | e135f0742b5e4130ae7c043c59ae32ca | /downloads/e135f0742b5e4130ae7c043c59ae32ca/japasen566761_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                               |
| 2018-09-19 17:25:02.234416 | 2019-12-09 11:35:40.343913 | 06f6156b32504d02ad1f904c1fc97e6e | /downloads/06f6156b32504d02ad1f904c1fc97e6e/jtiwari777973_course-v1:Myacademysitel+OLT_8190+2017_T2_Certificate.pdf                                                               |
| 2019-01-02 21:01:53.185314 | 2020-02-12 15:03:25.858944 | 1e2eed565d4c4ad79aa6d769e466070c | https://lt-learning-myacademy-default.s3.amazonaws.com/downloads/1e2eed565d4c4ad79aa6d769e466070c/jcabatcan818600_course-v1-Myacademysitel%2BKRO_0001%2B2017_T4_Certificate.pdf   |
| 2019-01-03 13:10:15.952684 | 2020-02-12 15:20:11.839642 | 02a1896cf0044220a21895c97abeb120 | https://lt-learning-myacademy-default.s3.amazonaws.com/downloads/02a1896cf0044220a21895c97abeb120/pdalimot625824_course-v1-Myacademysitel%2BKRO_0001%2B2017_T4_Certificate.pdf    |
| 2019-01-03 18:26:44.821030 | 2020-02-12 15:08:37.136609 | b33c5ad6f5074faea8a4d56417b9c092 | https://lt-learning-myacademy-default.s3.amazonaws.com/downloads/b33c5ad6f5074faea8a4d56417b9c092/wjoung1084736_course-v1-Myacademysitel%2BKRO_0001%2B2017_T4_Certificate.pdf     |
| 2019-01-04 07:02:56.044063 | 2020-02-12 15:17:11.756982 | 9f3e9b9fe5194537b746927e8c6b9731 | https://lt-learning-myacademy-default.s3.amazonaws.com/downloads/9f3e9b9fe5194537b746927e8c6b9731/mkumar868116_course-v1-Myacademysitel%2BKRO_0001%2B2017_T4_Certificate.pdf      |
| 2019-01-04 07:19:26.900076 | 2020-02-12 15:00:10.883794 | 3556773e4aca4187b924a753371d56f9 | https://lt-learning-myacademy-default.s3.amazonaws.com/downloads/3556773e4aca4187b924a753371d56f9/ctrinidad924905_course-v1-Myacademysitel%2BKRO_0001%2B2017_T4_Certificate.pdf   |
......
30 rows in set (0.00 sec)
```

# How we deploy it:
 - Command: `git pull .....` in folder `/edx/app/certs/certificate`     ( 🆕  New Exporting way )
 - `sudo /edx/bin/supervisorctl restart certs`
 - Clean all local files in folders `downloads` + `certs` by command : `/edx/app/certs/venvs/certs/bin/python /edx/app/certs/certificates/cleaner.py --target_type=dfs --dryrun=true  --start_date=2024-11-16` 👇 
```
cd /edx/app/certs
sudo -H -u www-data bash                       < ---------------------- User:    www-data
source /edx/app/certs/venvs/certs/bin/activate
```
 - Clean S3 PNGs of Learner certificate by command : `/edx/app/certs/venvs/certs/bin/python /edx/app/certs/certificates/cleaner.py --target_type=s3 --dryrun=true` 👇 
```
cd /edx/app/certs
sudo -H -u certs bash
source /edx/app/certs/venvs/certs/bin/activate
```

 - Tested on STAGE CN:
   -  Export certificates as zip file ( https://stage2.learning-tribes.com/courses/course-v1:edX+AGT101+2020_T1/instructor#view-certificates ) 👇 :
![image](https://github.com/user-attachments/assets/e2e6f0e9-4406-4441-9b7d-ebd9849396f6)
   - Logs: 
```
2024-11-19 09:27:18,297 INFO 18780 - [INFO] Downloading downloads/a199e5f2619e4df79f31f442920f7f6d/LT-developer_course-v1-edX+AGT101+2020_T1_Certificate.pdf to /tmp/certificates/nzwE9O/downloads/certs-zip/LT-developer_course-v1-edX+AGT101+2020_T1_Certificate.pdf...
2024-11-19 09:27:18,353 INFO 18780 - compressed a199e5f2619e4df79f31f442920f7f6d/LT-developer_course-v1-edX+AGT101+2020_T1_Certificate.pdf to /tmp/certificates/nzwE9O/downloads/certs-zip/course-v1-edX+AGT101+2020_T1_2024-11-19 09-27-18.257971.zip
2024-11-19 09:27:18,389 INFO 18780 - uploaded /tmp/certificates/nzwE9O/downloads/certs-zip/course-v1-edX+AGT101+2020_T1_2024-11-19 09-27-18.257971.zip to downloads/certs-zip/course-v1-edX+AGT101+2020_T1_2024-11-19 09-27-18.257971.zip
2024-11-19 09:27:18,389 INFO 18780 - [INFO] Removing tmp zip files folder: /tmp/certificates/nzwE9O/downloads/certs-zip
2024-11-19 09:27:18,389 INFO 18780 - [INFO] Closing AWS/S3 handle...
2024-11-19 09:27:18,389 INFO 18780 - Posting result to the LMS: {'xqueue_header': '{"submission_id": 2120, "submission_key": "e50bbe37017c3ed909ea6657171dd6fe"}', 'xqueue_body': '{"action": "export", "course_id": "course-v1:edX+AGT101+2020_T1", "url": "https://lt-learning-customer2-default.s3.amazonaws.com/downloads/certs-zip/course-v1-edX%2BAGT101%2B2020_T1_2024-11-19%2009-27-18.257971.zip"}'}
```

Commit:
COPY_TO_WEB_ROOT=False

Commit:
fixed: generate PNG file for Example Certificate Only

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
updated

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
Fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
fixed

Commit:
testing

Commit:
testing

Commit:
testing

Commit:
test

Commit:
test

Commit:
fixed

Commit:
fixed

BarryPaneer commented on gen_cert.py:
```diff
@@ -812,8 +815,8 @@ def _generate_v1_certificate(
-        # Convert PDF into PNG
-        if self._render_pdf_to_image_fitz(filename) and download_url and download_url.strip():
+        # Convert PDF into PNG for Example Certificates Only ! ( `_is_example_certificate` = True )
+        if self._is_example_certificate == True and self._render_pdf_to_image_fitz(filename) and download_url and download_url.strip():
```
- Only generate `.PNG` file for Example Certificate. Flag `_is_example_certificate` is passed from Platform. [CODE](https://github.com/Learningtribes/platform/blob/d8142a402b0599dcffadd1c04e328c981a06e234/lms/djangoapps/certificates/queue.py#L642)

lpfaender replied:
`if self._is_example_certificate == True and …` => `if self._is_example_certificate and …`

BarryPaneer commented on settings.py:
```diff
@@ -100,7 +100,6 @@
     CERT_DOWNLOAD_URL = ENV_TOKENS.get('CERT_DOWNLOAD_URL', "")
     CERT_VERIFY_URL = ENV_TOKENS.get('CERT_VERIFY_URL', "")
     CERT_WEB_ROOT = ENV_TOKENS.get('CERT_WEB_ROOT', CERT_WEB_ROOT)
-    COPY_TO_WEB_ROOT = ENV_TOKENS.get('COPY_TO_WEB_ROOT', COPY_TO_WEB_ROOT)
```
- Removed loading from `json cfg / COPY_TO_WEB_ROOT`, now the default value is `False`.

lpfaender replied:
I prefer we keep it and remove `COPY_TO_WEB_ROOT` from env.json in a PR for _hawthorn_inventory_.

BarryPaneer commented on certificate_agent.py:
```diff
@@ -118,17 +103,15 @@ def main():
         else:
             try:
+                # We fetch this extra parameter to differentiate example certificates from learner certificates.
+                is_example_certificate = xqueue_body.get('example_certificate', False)
```
- `True`: example certificate flag

BarryPaneer commented on certificate_agent.py:
```diff
@@ -71,41 +71,26 @@ def main():
-                download_url = export.create_and_upload(certs_path)
+                with CertificateExport(course_id, certs_path) as export:
+                    log.info('Generating Certificates Export for {}. Certs_path {}'.format(course_id, certs_path))
+                    download_url = export.create_and_upload()
```
- download s3 PDFs to tmp folder > zip them > upload new zip file to S3.

Commit:
added: try...catch...

Commit:
removed: one of retries

Commit:
fixed

BarryPaneer commented on cleaner.py:
```diff
@@ -0,0 +1,192 @@
+
+        ########### Start to run cleaning task ###########
+        _cleaner = DFSCleaner(args.dryrun) if args.target_type == 'dfs' else S3LearnerCertPNGCleaner(args.dryrun)
+        _cleaner.run()
```
- `/edx/app/certs/venvs/certs/bin/python /edx/app/certs/certificates/cleaner.py --target_type=[s3|dfs] --dryrun=[true|false]`

lpfaender commented on gen_cert.py:
```diff
@@ -230,6 +232,7 @@ def __init__(self, course_id, template_pdf=None, aws_id=None, aws_key=None, dir_
         self.score = 0
         self.duration = None
         self.completion = None
+        self._is_example_certificate = is_example_certificate
```
Let's remove the `_`: `self.is_example_certificate = is_example_certificate`.

lpfaender commented on gen_cert.py:
```diff
@@ -2181,69 +2184,118 @@ def fontlist_with_style(a_style):
+        self._ensure_dir(TMP_GEN_DIR)
+        self._dir_prefix = tempfile.mkdtemp(prefix=TMP_GEN_DIR)
+        self._ensure_dir(self._dir_prefix)
+        self._course_id = course_id
```
Let's remove the `_`.

Commit:
removed: _

Commit:
removed: _

Commit:
removed: _

Commit:
removed: _

lpfaender commented on gen_cert.py:
```diff
@@ -2181,69 +2184,118 @@ def fontlist_with_style(a_style):
+        self.initialize_s3_handles()
+        self.compress_s3_PDFs()
+
+        return self.upload_zip_file_to_s3()
```
I don't see where we delete the local files after uploading to S3.

BarryPaneer replied:
In `def __exit__()`

Commit:
supported: entering start date for cleaning tool

Commit:
reverted: as we keep loading COPY_TO_WEB_ROOT from json (value: False)

lpfaender commented on cleaner.py:
```diff
@@ -0,0 +1,194 @@
+import settings
+
+
+class _CleanerInterface(object):
```
I don't understand what this class brings.

BarryPaneer replied:
- Regulate the subclass to implement the method `def run(self):`.

lpfaender commented on cleaner.py:
```diff
@@ -0,0 +1,194 @@
+            raise Exception('[Error] Invalid target type: {}'.format(args.target_type))
+
+        ########### Start to run cleaning task ###########
+        cleaner = DFSCleaner(args.dryrun) if args.target_type == 'dfs' else S3LearnerCertPNGCleaner(args.dryrun)
```
Considering `target_type` expects `dfs` or `s3`, we should check the value:
```
if args.target_type == "dfs":
    …
elif args.target_type == "s3":
    …
else:
    …(error)
```

Or you replace `target_type` with a boolean argument `--s3`. If we don't have the `s3`, we assume it's local.

BarryPaneer replied:
```
        if args.target_type not in ('dfs', 's3'):
            raise Exception('[Error] Invalid target type: {}'.format(args.target_type))
```
 - actually we have a check. Let me refactor it.

lpfaender commented on cleaner.py:
```diff
@@ -0,0 +1,194 @@
+
+        ########### Start to run cleaning task ###########
+        cleaner = DFSCleaner(args.dryrun) if args.target_type == 'dfs' else S3LearnerCertPNGCleaner(args.dryrun)
+        cleaner.run()
```
I would pass `args.dryrun` in the `run` function instead of the constructor because there would be no reason to re-initialize the cleaner to change the type of run.
`clearner.run(args.dryrun)`

lpfaender commented on cleaner.py:
```diff
@@ -0,0 +1,194 @@
+        )
+
+    def run(self):
+        self.delete_png_once()
```
Considering we do nothing but call this function, there's no reason to create a separate function. You can move the `delete_png_once`  code directly inside `run.`

lpfaender commented on cleaner.py:
```diff
@@ -0,0 +1,194 @@
+        marker = None       # Used for pagination
+
+        while True:
+            is_printed = False
```
I don't understand what you do with `is_printed`… We already print after we've finished handling the batch of results.

BarryPaneer replied:
For debug only. If there's a large amount of s3 file which not needed to be removed. I don't why not message printed on screen. HTTP got stuck or still working.
I remove it. useless now.

lpfaender commented on cleaner.py:
```diff
@@ -0,0 +1,194 @@
+            except ValueError:
+                return True         # (Username): It's a Learner Certificate if ValueError raised here.
+
+        return None
```
We should avoid situations where we must separate True/False/None. It's not a bad idea to track unrecognized values, but then, I would suggest raising an exception when we find something that doesn't match the expected pattern of a username or UUID. You can easily catch the exception:

```
try:
    is_leaner_certificate = self.is_leaner_certificate(png_resource_uri)
    if is_leaner_certificate:
        print('[INFO] DELETING Learner Certificate PNG: {} '.format(png_resource_uri.encode('utf-8')))
        removed_learner_png_number += 1
        if not dryrun:
            self.bucket.delete_key(png_resource_uri)    
    else:
        example_cert_number += 1
except Exception:
     print('[ERROR] Got an Unrecognized URI : {}'.format(png_resource_uri.encode('utf-8')))
     unrecognized_number += 1
```

lpfaender commented on cleaner.py:
```diff
@@ -0,0 +1,194 @@
+    PERIOD_END_DATE = None
+
+    def __init__(self, dryrun=True):
+        self.PERIOD_END_DATE = datetime.now() - relativedelta(
```
I would prefer to enter a _YYYY-MM-DD_ date like "2019-12-06" and we remove everything that is ≥ to this date. We don't need a range.
As for `dryrun`, it would be better to set the argument in the command line and then pass it to `cleaner.run()`.

lpfaender commented on cleaner.py:
```diff
@@ -0,0 +1,194 @@
+    def run(self):
+        for _root_folder in self.TARGET_ROOT_FOLDERS:
+            print('[INFO] ################# Root folder {} #################'.format(_root_folder))
+            self.delete_resources_in_range(_root_folder)
```
Even though there's a bit more going on here than in `S3LearnerCertPNGCleaner`, we don't really need to create a separate function.

Commit:
fixed

Commit:
fixed

BarryPaneer commented:
@lpfaender  Fixed above.

```
certs@learning-tribes:~$ /edx/app/certs/venvs/certs/bin/python /edx/app/certs/certificates/cleaner.py --target_type=dfs --dryrun=true
[INFO] Dryrun mode : ON
Please enter Start date [Format: 2019-12-06]: 2024-11-15
[INFO] cleaning "DFS" files since 2024-11-15 00:00:00 in folders : /edx/var/certs/www-data/downloads + /edx/var/certs/www-data/cert
[INFO] ################# Root folder /edx/var/certs/www-data/downloads #################
......
Done !
certs@learning-tribes:~$ /edx/app/certs/venvs/certs/bin/python /edx/app/certs/certificates/cleaner.py --target_type=s3 --dryrun=true
[INFO] Dryrun mode : ON
[INFO] cleaning "AWS/S3" learner certificates ".PNG" files in Bucket[lt-learning-customer2-default]
[INFO] Example Certificate Number = 14, Unrecognized URI Number = 0, Removed Learner Certificate PNG Number = 0
Done !
```

Commit:
fixed

Commit:
fixed

BarryPaneer commented:
Fixed all. 

```
(certs) certs@learning-tribes:~/certificates$ /edx/app/certs/venvs/certs/bin/python /edx/app/certs/certificates/cleaner.py --target_type=dfs --dryrun=true --start_date=2024-11-16
[INFO] Dryrun mode : ON
[INFO] cleaning "DFS" files since 2024-11-16 00:00:00 in folders : /edx/var/certs/www-data/downloads + /edx/var/certs/www-data/cert
[INFO] ################# Root folder /edx/var/certs/www-data/downloads #################
[INFO] ################# Root folder /edx/var/certs/www-data/cert #################
Done !
(certs) certs@learning-tribes:~/certificates$ /edx/app/certs/venvs/certs/bin/python /edx/app/certs/certificates/cleaner.py --target_type=s3 --dryrun=true
[INFO] Dryrun mode : ON
[INFO] cleaning 'AWS/S3' learner certificates '.PNG' files in Bucket[lt-learning-customer2-default]
[INFO] Example Certificate Number = 14, Unrecognized URI Number = 0, Removed Learner Certificate PNG Number = 0
Done !
```
@lpfaender

Commit:
added: comments

Commit:
removed: unused

Commit:
added

Commit:
fixed

Commit:
testing

Commit:
fixed

BarryPaneer commented:
done.

```
certs@learning-tribes:~/certificates$ /edx/app/certs/venvs/certs/bin/python /edx/app/certs/certificates/cleaner.py --target_type=dfs --start_date=2024-11-16 --dryrun
[INFO] Dryrun mode : ON
[INFO] cleaning "DFS" files since 2024-11-16 00:00:00 in folders : /edx/var/certs/www-data/downloads + /edx/var/certs/www-data/cert
[INFO] ################# Root folder /edx/var/certs/www-data/downloads #################
[INFO] ################# Root folder /edx/var/certs/www-data/cert #################
Done !
certs@learning-tribes:~/certificates$ /edx/app/certs/venvs/certs/bin/python /edx/app/certs/certificates/cleaner.py --target_type=s3 --dryrun
[INFO] Dryrun mode : ON
[INFO] cleaning 'AWS/S3' learner certificates '.PNG' files in Bucket[lt-learning-customer2-default]
[INFO] Example Certificate Number = 14, Unrecognized URI Number = 0, Removed Learner Certificate PNG Number = 0
Done !
```

Commit:
fixed
